clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -313,6 +317,13 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
                 config.ignore(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);
                 this.valueSerializer = valueSerializer;
             }
+
+            // load interceptors and make sure they get clientId
+            userProvidedConfigs.put(ProducerConfig.CLIENT_ID_CONFIG, clientId);
+            List<ProducerInterceptor<K, V>> interceptorList = (List) (new ProducerConfig(userProvidedConfigs)).getConfiguredInstances(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG,
+                    ProducerInterceptor.class);
+            this.interceptors = new ProducerInterceptors<>(interceptorList);
```

Comment:
This might count in the micro-optimization category, but if there are no interceptors (which will probably be a majority of the time), maybe it would be better to store null and skip the interceptor invocation? This is less of a concern for the consumer since records are handled in batches, but the producer interceptor is called on each record pushed.